src/name/admitriev/jhelper/actions/ArchiveTaskAction.java

```diff
@@ -0,0 +1,186 @@
+            throw new NotificationException("No project found", "Are you in any project?");
+        }
+        VirtualFile file = e.getDataContext().getData(CommonDataKeys.VIRTUAL_FILE);
+        if (file == null) {
+            throw new NotificationException("No task or directory with tasks selected", "To archive a solution or a directory you should select it first");
+        }
+        VirtualFile projectBaseDir = ProjectUtil.guessProjectDir(project);
+        if (projectBaseDir == null) {
+            throw new NotificationException("Unable to find base directory of project", "Make sure you are not in default project");
+        }
+        Configurator configurator = project.getService(Configurator.class);
+        String archiveDirectory = configurator.getState().getArchiveDirectory();
+        ApplicationManager.getApplication().runWriteAction(new RecursiveArchiver(project, projectBaseDir, archiveDirectory, file));
+    }
+
+    private static final class RecursiveArchiver implements Runnable {
+        final Project project;
+        final VirtualFile projectBaseDir;
+        final String archiveDirectory;
+        final VirtualFile selectedFile;
+        final ArrayList<String> pathToFile;
+        final String curDate;
+        int archivedCount = 0;
+
+        public RecursiveArchiver(@NotNull Project project, @NotNull VirtualFile projectBaseDir, @NotNull String archiveDirectoryRelativePath, @NotNull VirtualFile selectedFile) {
+            this.project = project;
+            this.projectBaseDir = projectBaseDir;
+            this.archiveDirectory = archiveDirectoryRelativePath;
+            this.selectedFile = selectedFile;
+            this.pathToFile = new ArrayList<>();
+            this.curDate = DateTimeFormatter.ofPattern("dd_MM_yyyy_HH:mm:ss").format(LocalDateTime.now());
+        }
+
+        @Override
+        public void run() {
+            try {
+                VfsUtil.createDirectoryIfMissing(archiveDirectory);
+            } catch (IOException ex) {
+                throw new NotificationException("Unable to create archive directory", "Root of archive directory does not exist and failed to be created");
+            }
+            VfsUtilCore.visitChildrenRecursively(selectedFile, new VirtualFileVisitor<>() {
+                @Override
+                public boolean visitFile(@NotNull VirtualFile file) {
+                    if (file.isDirectory()) {
+                        pathToFile.add(file.getName());
+                        return true;
+                    }
+                    handleSourceFile(file);
+                    return false;
+                }
+
+                @Override
+                public void afterChildrenVisited(@NotNull VirtualFile file) {
+                    pathToFile.remove(pathToFile.size() - 1);
+                }
+            });
+            if (archivedCount > 0) {
+                LocalFileSystem.getInstance().refresh(true); // to notify Vfs about new files created in archive
```

Comment:
should we notify fs if something was created even if there was exception in the middle?